lib/cosmos/streams/tcpip_client_stream.rb

```diff
@@ -52,26 +52,70 @@ def initialize(hostname, write_port, read_port, write_timeout, read_timeout)
       end
 
       write_socket = nil
-      if write_addr
+      if @write_addr
         write_socket = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM, 0)
         write_socket.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1)
-        write_socket.connect(write_addr)
       end
 
       read_socket = nil
-      if read_addr
+      if @read_addr
         if @write_port != @read_port
           read_socket = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM, 0)
           read_socket.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1)
-          read_socket.connect(read_addr)
         else
           read_socket = write_socket
         end
       end
 
+      @connect_timeout = ConfigParser.handle_nil(connect_timeout)
+      @connect_timeout = @connect_timeout.to_f if @connect_timeout
+
       super(write_socket, read_socket, write_timeout, read_timeout)
     end
 
+    # Connect the socket(s)
+    def connect
+      if @write_socket
+        begin
+          @write_socket.connect_nonblock(@write_addr)
+        rescue IO::WaitWritable
+          begin
+            _, sockets, _ = IO.select(nil, [@write_socket], nil, @connect_timeout) # wait 3-way handshake completion
+          rescue Errno::ENOTSOCK
+            raise "Connect canceled"
+          end
+          if sockets and !sockets.empty?
+            begin
+              @write_socket.connect_nonblock(@write_addr) # check connection failure
+            rescue Errno::EISCONN
+            end
+          else
+            raise "Connect timeout"
+          end
+        end
+      end
+      if @read_socket and @read_socket != @write_socket
+        begin
+          @read_socket.connect_nonblock(@read_addr)
+        rescue IO::WaitWritable
```

Comment:
I'm trying to keep json_drb as independent of COSMOS as possible.  I'll refactor tcpip_client_stream though.